mp4parse/src/lib.rs

```diff
@@ -916,6 +1227,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+                    let mut data = vec_with_capacity(b.bytes_left().try_into()?)?;
+                    b.read_to_end(&mut data)?;
+                    vec_push(&mut mdats, MediaDataBox { offset, data })?;
+                }
+            }
+            _ => skip_box_content(&mut b)?,
+        }
+
+        check_parser_state!(b.content);
+    }
+
+    // If the `mdat` box came before the `meta` box, we need to fill in our primary item data
+    let primary_item_extents = primary_item_extents.ok_or(Error::InvalidData("primary item extents missing"))?;
+    for (extent, data) in primary_item_extents.iter().zip(primary_item_extents_data.iter_mut()) {
+        if data.is_empty()  {
+            // try to find an overlapping mdat
+            for mdat in &mut mdats {
+                if mdat.matches_extent(&extent.extent_range) {
+                    data.append(&mut mdat.data)
+                } else if mdat.contains_extent(&extent.extent_range) {
+                    mdat.read_extent(&extent.extent_range, data)?;
+                }
+            }
+        }
+    }
+
+    context.primary_item = primary_item_extents_data.into_iter().flatten().collect();
+
+    Ok(())
+}
+
+/// Parse a metadata box in the context of an AVIF
+/// Currently requires the primary item to be an av01 item type and generates
+/// an error otherwise.
+/// See ISO 14496-12:2015 § 8.11.1
+fn read_avif_meta<T: Read + Offset>(src: &mut BMFFBox<T>) -> Result<ItemLocationBoxItem> {
+    let version = read_fullbox_version_no_flags(src)?;
+
+    if version != 0 {
+        return Err(Error::Unsupported("unsupported meta version"))
+    }
+
+    let mut primary_item_id = None;
+    let mut item_infos = None;
+    let mut iloc_items = None;
+
+    let mut iter = src.box_iter();
+    while let Some(mut b) = iter.next_box()? {
+        match b.head.name {
+            BoxType::ItemInfoBox => {
+                if item_infos.is_some() {
+                    return Err(Error::InvalidData("There should be zero or one iinf boxes per ISO 14496-12:2015 § 8.11.6.1"));
+                }
+                item_infos = Some(read_iinf(&mut b)?);
+            }
+            BoxType::ItemLocationBox => {
+                if iloc_items.is_some() {
+                    return Err(Error::InvalidData("There should be zero or one iloc boxes per ISO 14496-12:2015 § 8.11.3.1"));
+                }
+                iloc_items = Some(read_iloc(&mut b)?);
+            }
+            BoxType::PrimaryItemBox => {
+                if primary_item_id.is_some() {
+                    return Err(Error::InvalidData("There should be zero or one iloc boxes per ISO 14496-12:2015 § 8.11.4.1"));
+                }
+                primary_item_id = Some(read_pitm(&mut b)?);
+            }
+            _ => skip_box_content(&mut b)?,
+        }
+
+        check_parser_state!(b.content);
+    }
+
+    let primary_item_id = primary_item_id.ok_or(Error::InvalidData("primary_item_id not present in iinf box"))?;
```

Comment:
Is this error case triggered by the `pitm` box being absent? If so, could we change the error string?

Reply:
Good catch! Will change to `"Required pitm box not present in meta box"`.